File: app/src/main/java/kr/co/yogiyo/rookiephotoapp/settings/SettingsFragment.java
package kr.co.yogiyo.rookiephotoapp.settings;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceFragmentCompat;

import kr.co.yogiyo.rookiephotoapp.GlobalApplication;
import kr.co.yogiyo.rookiephotoapp.R;

// TODO : preference엔 databinding이 지원되지 않는다고 해서 observe로 구현해야 할 것 같음
public class SettingsFragment extends PreferenceFragmentCompat implements Preference.OnPreferenceClickListener {

    private Preference backupDialogPreference;
    private Preference restoreDialogPreference;
    private Preference signDialogPreference;

    public static SettingsFragment newInstance() {
        return new SettingsFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public void onCreatePreferences(Bundle bundle, String rootKey) {
        setPreferencesFromResource(R.xml.preference_settings, rootKey);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        signDialogPreference = findPreference(SettingsActivity.SIGN_DIALOG_KEY);
        backupDialogPreference = findPreference(SettingsActivity.BACKUP_DIALOG_KEY);
        restoreDialogPreference = findPreference(SettingsActivity.RESTORE_DIALOG_KEY);

        if (GlobalApplication.getGlobalApplicationContext().getFirebaseAuth().getCurrentUser() == null) {
            signDialogPreference.setTitle(getString(R.string.text_need_to_signin));
        } else {
            signDialogPreference.setTitle(GlobalApplication.getGlobalApplicationContext().getFirebaseAuth().getCurrentUser().getEmail());
        }
    }

    @Override
    public void onDisplayPreferenceDialog(Preference preference) {
        DialogFragment dialogFragment = null;
        if (preference instanceof SettingsDialogPreference) {
            if (GlobalApplication.getGlobalApplicationContext().getFirebaseAuth().getCurrentUser() == null) {
                preference = signDialogPreference;
                ((SettingsDialogPreference) preference).setDialogLayoutResource(R.layout.dialog_fragment_sign);
                dialogFragment = SignDialogFragment.newInstance();
            } else {
                switch (preference.getKey()) {
                    case SettingsActivity.SIGN_DIALOG_KEY:
                        ((SettingsDialogPreference) preference).setDialogLayoutResource(R.layout.dialog_fragment_signout);
                        dialogFragment = SignOutDialogFragment.newInstance();
                        break;
                    case SettingsActivity.BACKUP_DIALOG_KEY:
                        ((SettingsDialogPreference) preference).setDialogLayoutResource(R.layout.dialog_fragment_backup_restore);
                        dialogFragment = BackupRestoreDialogFragment.newInstance();
                        break;
                    case SettingsActivity.RESTORE_DIALOG_KEY:
                        ((SettingsDialogPreference) preference).setDialogLayoutResource(R.layout.dialog_fragment_backup_restore);
                        dialogFragment = BackupRestoreDialogFragment.newInstance();
                        break;
                }
            }

            Bundle bundle = new Bundle(1);
            bundle.putString(SettingsActivity.PREFERENCE_KEY, preference.getKey());
            dialogFragment.setArguments(bundle);
        }

        if (dialogFragment != null && getFragmentManager() != null) {
            dialogFragment.setTargetFragment(this, 0);
            dialogFragment.show(getFragmentManager(), SettingsActivity.PREFERENCE_DIALOG_TAG);
        } else {
            super.onDisplayPreferenceDialog(preference);
        }
    }

    @Override
    public boolean onPreferenceClick(Preference preference) {
        switch (preference.getKey()) {
            case SettingsActivity.BACKUP_DIALOG_KEY:
                onDisplayPreferenceDialog(backupDialogPreference);
                return true;
            case SettingsActivity.RESTORE_DIALOG_KEY:
                onDisplayPreferenceDialog(restoreDialogPreference);
                return true;
            default:
                return false;
        }
    }
}

File: app/src/main/java/kr/co/yogiyo/rookiephotoapp/ui/setting/SettingsFragment.kt
package kr.co.yogiyo.rookiephotoapp.ui.setting

import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v7.preference.Preference
import android.support.v7.preference.PreferenceFragmentCompat

import kr.co.yogiyo.rookiephotoapp.GlobalApplication
import kr.co.yogiyo.rookiephotoapp.R
import kr.co.yogiyo.rookiephotoapp.ui.setting.backup_restore.BackupRestoreDialogFragment
import kr.co.yogiyo.rookiephotoapp.ui.setting.auth.SignDialogFragment
import kr.co.yogiyo.rookiephotoapp.ui.setting.auth.SignOutDialogFragment

// TODO : preference엔 databinding이 지원되지 않는다고 해서 observe로 구현해
class SettingsFragment : PreferenceFragmentCompat(), Preference.OnPreferenceClickListener {

    private val backupDialogPreference by lazy { findPreference(SettingsActivity.BACKUP_DIALOG_KEY) }
    private val restoreDialogPreference by lazy { findPreference(SettingsActivity.RESTORE_DIALOG_KEY) }
    private val signDialogPreference by lazy { findPreference(SettingsActivity.SIGN_DIALOG_KEY) }
    private val signOutDialogFragment by lazy {
        SignOutDialogFragment.newInstance().apply {
            signOut = {
                try {
                    GlobalApplication.globalApplicationContext.firebaseAuth.signOut()
                    signDialogPreference.title = getString(R.string.text_need_to_signin)
                    this.onSuccess()
                } catch (e: Exception) {
                    e.printStackTrace()
                    this.onFail()
                }
            }
        }
    }
    private val signDialogFragment by lazy {
        SignDialogFragment.newInstance().apply {
            signInWithEmailAndPassword = { email, password ->
                if (email.isEmpty() || password.isEmpty()) {
                    this.onFail()
                } else {
                    GlobalApplication.globalApplicationContext.firebaseAuth.signInWithEmailAndPassword(email, password)
                            .addOnCompleteListener { task ->
                                if (task.isSuccessful) {
                                    this.onSuccess()
                                    (context as SettingsActivity).showToast(R.string.text_signin_success)
                                    signDialogPreference.title = GlobalApplication.globalApplicationContext.firebaseAuth.currentUser?.email
                                            ?: getString(R.string.text_need_to_signin)
                                } else {
                                    this.onFail()
                                }
                            }
                }
            }
            createUserWithEmailAndPassword = { email, password ->
                if (email.isEmpty() || password.isEmpty()) {
                    this.onFail()
                } else {
                    GlobalApplication.globalApplicationContext.firebaseAuth.createUserWithEmailAndPassword(email, password)
                            .addOnCompleteListener { task ->
                                if (task.isSuccessful) {
                                    this.onSuccess()
                                    (context as SettingsActivity).showToast(R.string.text_signup_success)
                                    signDialogPreference.title = GlobalApplication.globalApplicationContext.firebaseAuth.currentUser?.email
                                            ?: getString(R.string.text_need_to_signin)
                                } else {
                                    this.onFail()
                                }
                            }
                }
            }
        }
    }

    private lateinit var dialogFragment: DialogFragment

    override fun onCreatePreferences(bundle: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preference_settings, rootKey)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        signDialogPreference.title = GlobalApplication.globalApplicationContext.firebaseAuth.currentUser?.email
                ?: getString(R.string.text_need_to_signin)
    }

    override fun onDisplayPreferenceDialog(preference: Preference?) {
        if (preference is SettingsDialogPreference) {
            if (GlobalApplication.globalApplicationContext.firebaseAuth.currentUser == null) {
                preference.dialogLayoutResource = R.layout.dialog_fragment_sign
                dialogFragment = signDialogFragment
            } else {
                preference.dialogLayoutResource = 0
                when (preference.key) {
                    SettingsActivity.SIGN_DIALOG_KEY -> { dialogFragment = signOutDialogFragment }
                    SettingsActivity.BACKUP_DIALOG_KEY, SettingsActivity.RESTORE_DIALOG_KEY -> {
                        dialogFragment = BackupRestoreDialogFragment.newInstance()
                    }
                }
            }

            dialogFragment.arguments = Bundle(1).apply {
                putString(SettingsActivity.PREFERENCE_KEY, preference.key)
            }
        }

        fragmentManager?.let { fragmentManager ->
            dialogFragment.run {
                setTargetFragment(this@SettingsFragment, 0)
                show(fragmentManager, SettingsActivity.PREFERENCE_DIALOG_TAG)
            }
        } ?: super.onDisplayPreferenceDialog(preference)
    }

    override fun onPreferenceClick(preference: Preference): Boolean {
        return when (preference.key) {
            SettingsActivity.BACKUP_DIALOG_KEY -> {
                GlobalApplication.globalApplicationContext.firebaseAuth.currentUser?.let {
                    onDisplayPreferenceDialog(backupDialogPreference)
                } ?: onDisplayPreferenceDialog(signDialogPreference)
                true
            }
            SettingsActivity.RESTORE_DIALOG_KEY -> {
                GlobalApplication.globalApplicationContext.firebaseAuth.currentUser?.let {
                    onDisplayPreferenceDialog(restoreDialogPreference)
                } ?: onDisplayPreferenceDialog(signDialogPreference)
                true
            }
            else -> false
        }
    }

    companion object {

        fun newInstance(): SettingsFragment {
            return SettingsFragment()
        }
    }
}


